persist/fs/commitlog/commit_log.go

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package commitlog
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3db/interfaces/m3db"
+	"github.com/m3db/m3x/metrics"
+	"github.com/m3db/m3x/time"
+)
+
+const (
+	commitLogWriteQueue = 32768 // at 1MM write/s allow to fall behind by ~6.5% our deadline
+)
+
+var (
+	errCommitLogClosed = errors.New("commit log is closed")
+
+	writeRetries    = 2
+	writeRetryDelay = time.Millisecond
+	timeNow         = time.Now
+	timeZero        = time.Time{}
+)
+
+type completionFn func(err error)
+
+type commitLog struct {
+	sync.RWMutex
+	opts    m3db.DatabaseOptions
+	metrics xmetrics.Scope
+	nowFn   m3db.NowFn
+	// TODO(r): replace buffered channel with concurrent striped
+	// circular buffer to avoid central write lock contention
+	writes         chan commitLogWrite
+	writer         commitLogWriter
+	pendingAcks    []completionFn
+	bitset         bitset
+	writerExpireAt time.Time
+	closed         bool
+	closeErr       chan error
+}
+
+type commitLogWrite struct {
+	series       m3db.CommitLogSeries
+	datapoint    m3db.Datapoint
+	unit         xtime.Unit
+	annotation   []byte
```

Comment:
Sure, we should probably do this across everything in one go too soon as I know plenty of other places where it's not.